### PR TITLE
CLOUDSTACK-8826: XenServer - Use device id passed as part of attach v…

### PR DESCRIPTION
**Pull request by koushik-das**
…olume API properly

If device id passed as part of API and available then use it otherwise fallback on XS to automatically assign one.
For ISO device id used is 3 and it is processed before any other entry to avoid conflict.

This is only specific to XS, ran the following tests:

BVT: test/integration/smoke/test_volumes.py

Test Volume creation for all Disk Offerings (incl. custom) ... === TestName: test_01_create_volume | Status : SUCCESS ===
ok
Attach a created Volume to a Running VM ... === TestName: test_02_attach_volume | Status : SUCCESS ===
ok
Download a Volume attached to a VM ... === TestName: test_03_download_attached_volume | Status : SUCCESS ===
ok
Delete a Volume attached to a VM ... === TestName: test_04_delete_attached_volume | Status : SUCCESS ===
ok
Detach a Volume attached to a VM ... === TestName: test_05_detach_volume | Status : SUCCESS ===
ok
Download a Volume unattached to an VM ... === TestName: test_06_download_detached_volume | Status : SUCCESS ===
ok
Test resize (negative) non-existent volume ... === TestName: test_07_resize_fail | Status : SUCCESS ===
ok
Test resize a volume ... === TestName: test_08_resize_volume | Status : SUCCESS ===
ok
Delete a Volume unattached to an VM ... === TestName: test_09_delete_detached_volume | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 9 tests in 1116.972s

OK

Regression: test/integration/component/test_volumes.py

Test Volume attach/detach to VM (5 data volumes) ... === TestName: test_01_volume_attach_detach | Status : SUCCESS ===
ok
Test Attach volumes (max capacity) ... === TestName: test_01_volume_attach | Status : SUCCESS ===
ok
Test attach volumes (more than max) to an instance ... === TestName: test_02_volume_attach_max | Status : SUCCESS ===
ok
Test Volumes and ISO attach ... === TestName: test_01_volume_iso_attach | Status : SUCCESS ===
ok
Test custom disk sizes beyond range ... === TestName: test_deployVmWithCustomDisk | Status : SUCCESS ===
ok
@Desc:Volume is not retaining same uuid when migrating from one ... SKIP: No suitable storage pools found for volume migration.                        Skipping
Attach a created Volume to a Running VM ... === TestName: test_01_attach_volume | Status : SUCCESS ===
ok
Detach a Volume attached to a VM ... === TestName: test_02_detach_volume | Status : SUCCESS ===
ok
Delete a Volume unattached to an VM ... === TestName: test_03_delete_detached_volume | Status : SUCCESS ===
ok
Create a volume under a non-root domain as non-root-domain user ... === TestName: test_create_volume_under_domain | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 10 tests in 1750.686s

OK (SKIP=1)

### CODE REVIEW

**Comment by koushik-das** [comment]
I had to undo changes from PR https://github.com/apache/cloudstack/pull/773 to run the tests.

**Comment by bhaisaab** [comment]
code LGTM

**Comment by koushik-das** [comment]
Can anyone review this?

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java [hunk 1]
```diff
@@ -3407,6 +3390,18 @@ public void handleSrAndVdiDetach(final String iqn, final Connection conn) throws
...
 
+    protected void destroyUnattachedVBD(Connection conn, VM vm) {
+        try {
+            for (VBD vbd : vm.getVBDs(conn)) {
+                if (Types.VbdType.DISK.equals(vbd.getType(conn)) && !vbd.getCurrentlyAttached(conn)) {
+                    vbd.destroy(conn);
+                }
+            }
+        } catch (final Exception e) {
+            s_logger.debug("destroyUnattachedVBD failed due to " + e.toString());
```
Comment (DaanHoogland) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java]: maybe add the exception or/and catch more specific exceptions

**Comment by DaanHoogland** [comment]
one remark, lgtm

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java (hunk 1 above)
Comment (koushik-das) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java]: @DaanHoogland Are you suggesting to log the full exception? Currently only the message is getting logged. The destroyUnattachedVBD is a best-effort call to cleanup VBDs.

Comment (DaanHoogland) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java]: @koushik-das I would catch only specific exceptions and give them specific messages and report on results in a finally block. the present state is that any numberformat, memory, stackoverflow or devidebyzero will be caught here, where it is not relevant. My initial remark was to make sure we know what kind of exception is caught at least. This is not always clear from the message which may be null.

Comment (koushik-das) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java]: @DaanHoogland Now logging the full exception

**Comment by koushik-das** [comment]
Merging as 2 LGTMs.

**Comment by pdion891** [comment]
@koushik-das  does this fix got tested with HVM vm having more than 4 VDI?  because we are experiencing issue where an HVM vm on XenServer 6.5 having 4 vdi (1 root + 3 datadisk) fail to start after a shutdown and it seams to be related to this part of code.  This is currently working on 4.4.x version of ACS.

**Comment by simongodard** [comment]
I confirm what @pdion891 described in the previous comment. 
http://markmail.org/thread/4nmyra6aofxtu3o2

**Comment by koushik-das** [comment]
@pdion891 This wasn't tested with HVM VMs.

**Pull request merged**